### ISSUE TITLE
[BUG] Need Cpp2's equivalent of Cpp1 `using base::member`

### ISSUE THREAD

**Issue by JohelEGP** [hsutter/cppfront#559]
<a id="title"/>

**[T](#title)itle**: Need Cpp2's equivalent of Cpp1 `using base::member`.

<a id="desc"/>

**[D](#desc)escription**:

I'm writing an iterator
and using [Boost.STLInterfaces][] to reduce boilerplate.
I implement `operator++: (inout this) -> forward iterator`,
but I need Cpp2's equivalent of Cpp1 `using base::member`
for the base `boost::stl_interfaces::iterator_interface`
to give me post-increment since my `operator++` hides it.
That's explained at ["An Important Note About operator++() and operator--()"](https://www.boost.org/doc/libs/release/doc/html/boost_stlinterfaces/tutorial___iterator_interface_.html#boost_stlinterfaces.tutorial___iterator_interface_.an_important_note_about__code__phrase_role__keyword__operator__phrase__phrase_role__special________phrase___code__and__code__phrase_role__keyword__operator__phrase__phrase_role__special________phrase___code_).

[Boost.STLInterfaces]: https://www.boost.org/doc/libs/release/doc/html/stl_interfaces.html

<a id="rep"/>
<a id="cpp2"/>

<details><summary>

**[M](#rep)inimal reproducer** (<https://cpp2.godbolt.org/z/xhsTbq8xq>):
</summary>

```Cpp2
#include <iterator>
#include <boost/stl_interfaces/iterator_interface.hpp>

sf: namespace = {
Vector2i: @value type = {
  public x: i32 = ();
  public y: i32 = ();
}
}

points_in_grid: @struct type = {
  grid_size: sf::Vector2i;

  iterator: @basic_value type = {
    this: boost::stl_interfaces::iterator_interface<
      iterator, std::forward_iterator_tag, sf::Vector2i, sf::Vector2i> = ();
    max: sf::Vector2i = ();
    pos: sf::Vector2i = ();

    operator=: (out this, grid_size: sf::Vector2i) = max = grid_size;

    operator*: (this) -> sf::Vector2i            = pos;
    operator++: (inout this) -> forward iterator = {
      [[assert: pos.y != max.y]]
      pos.x++;
      if pos.x == max.x {
        pos.x = 0;
        pos.y++;
      }
      return this;
    }
    operator==: (this, that) -> bool [[pre: max == that.max]] = pos == that.pos;
    operator==: (this, _: std::default_sentinel_t) -> bool    = {
      _ = _;
      if pos.y == max.y {
        [[assert: pos.x == 0]]
        return true;
      }
      return false;
    }
  }

  begin: (this) -> iterator              = iterator(grid_size);
  end: (this) -> std::default_sentinel_t = ();
}

static_assert(std::forward_iterator<points_in_grid::iterator>);

main: () = { }
```
</details>

<a id="cmd"/>
<details><summary>
<b><a href="#cmd">C</a>ommands</b>:
</summary>

```bash
cppfront main.cpp2
clang++17 -std=c++23 -stdlib=libc++ -lc++abi -pedantic-errors -Wall -Wextra -Wconversion -fsanitize=undefined -Werror=unused-result -I . main.cpp
```
</details>

<a id="expected"/>

**[E](#expected)xpected result**: A Cpp2's equivalent of Cpp1 `using base::member`.

<a id="actual"/>

**[A](#actual)ctual result and error**:

<a id="cpp1"/>
<details><summary>
<a href="#cpp1">C</a>pp2 lowered to Cpp1:
</summary>

```C++


//=== Cpp2 type declarations ====================================================


#include "cpp2util.h"


namespace sf {
class Vector2i;
  

}

class points_in_grid;
  

//=== Cpp2 type definitions and function declarations ===========================

#include <iterator>
#include <boost/stl_interfaces/iterator_interface.hpp>

namespace sf {
class Vector2i {
  public: cpp2::i32 x {}; 
  public: cpp2::i32 y {}; 
  public: [[nodiscard]] auto operator<=>(Vector2i const& that) const -> std::strong_ordering = default;
public: Vector2i(Vector2i const& that);

public: auto operator=(Vector2i const& that) -> Vector2i& ;
public: Vector2i(Vector2i&& that) noexcept;
public: auto operator=(Vector2i&& that) noexcept -> Vector2i& ;
public: explicit Vector2i();

};
}

class points_in_grid {
  public: sf::Vector2i grid_size; 

  public: class iterator: public boost::stl_interfaces::iterator_interface<iterator,std::forward_iterator_tag,sf::Vector2i,sf::Vector2i> {

    private: sf::Vector2i max {}; 
    private: sf::Vector2i pos {}; 

    public: explicit iterator(cpp2::in<sf::Vector2i> grid_size);

    public: [[nodiscard]] auto operator*() const -> sf::Vector2i;
    public: [[nodiscard]] auto operator++() -> iterator&;
      

    public: [[nodiscard]] auto operator==(iterator const& that) const -> bool;
    public: [[nodiscard]] auto operator==(cpp2::in<std::default_sentinel_t> _) const -> bool;
      
      public: iterator(iterator const& that);
public: explicit iterator();


  };

  public: [[nodiscard]] auto begin() const -> iterator;
  public: [[nodiscard]] auto end() const -> std::default_sentinel_t;
};

static_assert(std::forward_iterator<points_in_grid::iterator>);

auto main() -> int;


//=== Cpp2 function definitions =================================================


namespace sf {


Vector2i::Vector2i(Vector2i const& that)
                                : x{ that.x }
                                , y{ that.y }{}

auto Vector2i::operator=(Vector2i const& that) -> Vector2i& {
                                x = that.x;
                                y = that.y;
                                return *this;}
Vector2i::Vector2i(Vector2i&& that) noexcept
                                : x{ std::move(that).x }
                                , y{ std::move(that).y }{}
auto Vector2i::operator=(Vector2i&& that) noexcept -> Vector2i& {
                                x = std::move(that).x;
                                y = std::move(that).y;
                                return *this;}
Vector2i::Vector2i(){}
}

    points_in_grid::iterator::iterator(cpp2::in<sf::Vector2i> grid_size)
                                                     : boost::stl_interfaces::iterator_interface<iterator,std::forward_iterator_tag,sf::Vector2i,sf::Vector2i>{  }
                                                     , max{ grid_size }
     {  }

    [[nodiscard]] auto points_in_grid::iterator::operator*() const -> sf::Vector2i { return pos; }
    [[nodiscard]] auto points_in_grid::iterator::operator++() -> iterator&{
      cpp2::Default.expects(pos.y!=max.y, "");
      ++pos.x;
      if (pos.x==max.x) {
        pos.x = 0;
        ++pos.y;
      }
      return (*this); 
    }
    [[nodiscard]] auto points_in_grid::iterator::operator==(iterator const& that) const -> bool { 
                                                                cpp2::Default.expects(max==that.max, "");
    return pos==that.pos;  }
    [[nodiscard]] auto points_in_grid::iterator::operator==(cpp2::in<std::default_sentinel_t> _) const -> bool{
      (void) _;
      if (pos.y==max.y) {
        cpp2::Default.expects(pos.x==0, "");
        return true; 
      }
      return false; 
    }

    points_in_grid::iterator::iterator(iterator const& that)
                                : boost::stl_interfaces::iterator_interface<iterator,std::forward_iterator_tag,sf::Vector2i,sf::Vector2i>{ that }
                                , max{ that.max }
                                , pos{ that.pos }{}
points_in_grid::iterator::iterator()
                          : boost::stl_interfaces::iterator_interface<iterator,std::forward_iterator_tag,sf::Vector2i,sf::Vector2i>{  }{}

  [[nodiscard]] auto points_in_grid::begin() const -> iterator { return iterator(grid_size); }
  [[nodiscard]] auto points_in_grid::end() const -> std::default_sentinel_t { return {  };  }

auto main() -> int{}
```
</details>

<a id="out"/>
<details><summary>
<a href="#out">O</a>utput:
</summary>

```output
main.cpp2:47:15: error: static assertion failed
   47 | static_assert(std::forward_iterator<points_in_grid::iterator>);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
main.cpp2:47:15: note: because 'points_in_grid::iterator' does not satisfy 'forward_iterator'
/opt/compiler-explorer/clang-trunk-20230726/bin/../include/c++/v1/__iterator/concepts.h:156:3: note: because 'points_in_grid::iterator' does not satisfy 'input_iterator'
  156 |   input_iterator<_Ip> &&
      |   ^
/opt/compiler-explorer/clang-trunk-20230726/bin/../include/c++/v1/__iterator/concepts.h:139:3: note: because 'points_in_grid::iterator' does not satisfy 'input_or_output_iterator'
  139 |   input_or_output_iterator<_Ip> &&
      |   ^
/opt/compiler-explorer/clang-trunk-20230726/bin/../include/c++/v1/__iterator/concepts.h:115:3: note: because 'points_in_grid::iterator' does not satisfy 'weakly_incrementable'
  115 |   weakly_incrementable<_Ip>;
      |   ^
/opt/compiler-explorer/clang-trunk-20230726/bin/../include/c++/v1/__iterator/concepts.h:97:8: note: because '__i++' would be invalid: cannot increment value of type 'points_in_grid::iterator'
   97 |     __i++;                        // not required to be equality-preserving
      |        ^
1 error generated.
```
</details>

**Comment by JohelEGP** [comment]
See also:
> To enable authoring the `gram` from <https://github.com/hsutter/cppfront/issues/284#issuecomment-1478872698> as a library,
> we will need Cpp2's equivalent of Cpp1's `using namespace`.

**Comment by JohelEGP** [comment]
This also relates with
<https://github.com/hsutter/cppfront/issues/438#issuecomment-1538121203> (nested namespace definition) and
#467
in that they can declare a nested name (in Cpp1).
Although the Cpp2 `using` declaration might not use the Cpp2 declaration syntax.

**Comment by JohelEGP** [comment]
There's Cpp1's `using enum`, too.

It seems like Cpp2's `enum`s will be type metafunctions.
Those can't have an equivalent,
as Cpp1 doesn't have `using namespace` for class types.
Although it _might_ be possible to implement an equivalent in a Cppfront with [name lookup](https://github.com/hsutter/cppfront/commit/d8c1a50f22c1b171a50e87ccdb609fb05f41c021#commitcomment-111968233).

**Comment by farmerpiki** [comment]
a proposal, we could make the syntax
`: namespace == std::string_literals;`
equivalate to using namespace... without introducing new keywords ... because it's the same as aliasing it into the current namespace @JohelEGP ?
I guess the same syntax could work for multiple things than
I also ran into needing this today

**Comment by farmerpiki** [comment]
and we could also allow it inside a scope like a for or a inspect, by doing the same we do with local variables
(: namespace == aliased_namespace) for ...

**Comment by mrdomino** [comment]
Is this the same for e.g. `using std::swap;` in a function body? Would you say `: == std::swap;`?

**Comment by JohelEGP** [comment]
Resolved with commit dd047033de5a691721069eb9ddb5b4f7892afa94.
